In [1]:
# Import required libraries for hotel booking data analysis and preprocessing
import numpy as np
import pandas as pd

In [2]:
# Load the hotel booking dataset into a pandas DataFrame
df = pd.read_csv("hotel_bookings.csv")

In [3]:
# Display the first five rows of the dataset
df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


In [4]:
# Check the number of rows and columns in the dataset
df.shape

(119390, 32)

In [5]:
# Display dataset information, including data types and non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [6]:
# Check the number of missing values in each column
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [7]:
# Calculate the percentage of missing values in each column
missing_percentage = df.isnull().mean() * 100
missing_percentage

hotel                              0.000000
is_canceled                        0.000000
lead_time                          0.000000
arrival_date_year                  0.000000
arrival_date_month                 0.000000
arrival_date_week_number           0.000000
arrival_date_day_of_month          0.000000
stays_in_weekend_nights            0.000000
stays_in_week_nights               0.000000
adults                             0.000000
children                           0.003350
babies                             0.000000
meal                               0.000000
country                            0.408744
market_segment                     0.000000
distribution_channel               0.000000
is_repeated_guest                  0.000000
previous_cancellations             0.000000
previous_bookings_not_canceled     0.000000
reserved_room_type                 0.000000
assigned_room_type                 0.000000
booking_changes                    0.000000
deposit_type                    

In [8]:
# Display only columns with missing values
missing_percentage[missing_percentage > 0]

children     0.003350
country      0.408744
agent       13.686238
company     94.306893
dtype: float64

In [9]:
# Display the first 10 unique values in each column

print("Children:", df['children'].unique()[:10])
print("Country:", df['country'].unique()[:10])
print("Agent:", df['agent'].unique()[:10])
print("Company:", df['company'].unique()[:10])

Children: [ 0.  1.  2. 10.  3. nan]
Country: ['PRT' 'GBR' 'USA' 'ESP' 'IRL' 'FRA' nan 'ROU' 'NOR' 'OMN']
Agent: [ nan 304. 240. 303.  15. 241.   8. 250. 115.   5.]
Company: [ nan 110. 113. 270. 178. 240. 154. 144. 307. 268.]


In [10]:
# Count the number of unique values in each column
print("Children:", df['children'].nunique())
print("Country:", df['country'].nunique())
print("Agent:", df['agent'].nunique())
print("Company:", df['company'].nunique())

Children: 5
Country: 177
Agent: 333
Company: 352


In [11]:
# Count the number of bookings for each company
df['company'].value_counts()

company
40.0     927
223.0    784
67.0     267
45.0     250
153.0    215
        ... 
104.0      1
531.0      1
160.0      1
413.0      1
386.0      1
Name: count, Length: 352, dtype: int64

In [12]:
# Count bookings with a company ID
print("Non-missing:", df['company'].notna().sum())

# Count unique company IDs
print("Unique companies:", df['company'].nunique())

# Count bookings without a company ID
print("Missing:", df['company'].isna().sum())

Non-missing: 6797
Unique companies: 352
Missing: 112593


In [13]:
# Count bookings with and without an agent ID
print("Non-missing:", df['agent'].notna().sum())
print("Unique agents:", df['agent'].nunique())
print("Missing:", df['agent'].isna().sum())

Non-missing: 103050
Unique agents: 333
Missing: 16340


In [14]:
# Count bookings for each agent
df['agent'].value_counts().head(10)

agent
9.0      31961
240.0    13922
1.0       7191
14.0      3640
7.0       3539
6.0       3290
250.0     2870
241.0     1721
28.0      1666
8.0       1514
Name: count, dtype: int64

In [15]:
# Calculate the percentage of bookings for the top 10 agents
top_10_agents = df['agent'].value_counts().head(10)

top_10_agents_percentage = top_10_agents / df['agent'].notna().sum() * 100

top_10_agents_percentage

agent
9.0      31.015041
240.0    13.509947
1.0       6.978166
14.0      3.532266
7.0       3.434255
6.0       3.192625
250.0     2.785056
241.0     1.670063
28.0      1.616691
8.0       1.469190
Name: count, dtype: float64

In [16]:
# Count bookings for each market segment
df['market_segment'].value_counts()

market_segment
Online TA        56477
Offline TA/TO    24219
Groups           19811
Direct           12606
Corporate         5295
Complementary      743
Aviation           237
Undefined            2
Name: count, dtype: int64

In [17]:
# Calculate the percentage of bookings for each market segment
market_segment_percentage = df['market_segment'].value_counts(normalize=True) * 100

market_segment_percentage

market_segment
Online TA        47.304632
Offline TA/TO    20.285619
Groups           16.593517
Direct           10.558673
Corporate         4.435045
Complementary     0.622330
Aviation          0.198509
Undefined         0.001675
Name: proportion, dtype: float64

In [18]:
# Calculate the cancellation rate for each market segment
cancellation_by_segment = df.groupby('market_segment')['is_canceled'].mean() * 100

cancellation_by_segment.sort_values(ascending=False)

market_segment
Undefined        100.000000
Groups            61.062036
Online TA         36.721143
Offline TA/TO     34.316033
Aviation          21.940928
Corporate         18.734655
Direct            15.341901
Complementary     13.055182
Name: is_canceled, dtype: float64

In [19]:
# Compare booking volume and cancellation rate by market segment
segment_analysis = df.groupby('market_segment').agg(
    bookings=('is_canceled', 'count'),
    cancellation_rate=('is_canceled', 'mean')
)

segment_analysis['cancellation_rate'] *= 100

segment_analysis.sort_values('cancellation_rate', ascending=False)

,bookings,cancellation_rate
market_segment,,
Undefined,2,100.000000
Groups,19811,61.062036
Online TA,56477,36.721143
Offline TA/TO,24219,34.316033
Aviation,237,21.940928
Corporate,5295,18.734655
Direct,12606,15.341901
Complementary,743,13.055182


In [20]:
# Count canceled bookings for each market segment
canceled_by_segment = df.groupby('market_segment')['is_canceled'].sum()

canceled_by_segment.sort_values(ascending=False)

market_segment
Online TA        20739
Groups           12097
Offline TA/TO     8311
Direct            1934
Corporate          992
Complementary       97
Aviation            52
Undefined            2
Name: is_canceled, dtype: int64

In [21]:
# Summarize total, canceled, and remaining bookings by market segment
segment_summary = df.groupby('market_segment').agg(
    total_bookings=('is_canceled', 'count'),
    canceled_bookings=('is_canceled', 'sum')
)

# Calculate the number of non-canceled bookings
segment_summary['remaining_bookings'] = (
    segment_summary['total_bookings']
    - segment_summary['canceled_bookings']
)

# Sort by total bookings
segment_summary = segment_summary.sort_values(
    'total_bookings',
    ascending=False
)

segment_summary

,total_bookings,canceled_bookings,remaining_bookings
market_segment,,,
Online TA,56477,20739,35738
Offline TA/TO,24219,8311,15908
Groups,19811,12097,7714
Direct,12606,1934,10672
Corporate,5295,992,4303
Complementary,743,97,646
Aviation,237,52,185
Undefined,2,2,0


### Cancellation Analysis by Market Segment

The Groups segment has the highest meaningful cancellation rate at approximately 61.06%. Online TA has the largest number of canceled bookings, with 20,739 cancellations, partly reflecting its high booking volume. Direct bookings have a lower cancellation rate of approximately 15.34%.

These findings suggest that cancellation patterns differ across market segments. The Undefined segment has only two bookings, so its 100% cancellation rate should not be considered representative.

In [22]:
# Count bookings for each deposit type
df['deposit_type'].value_counts()

deposit_type
No Deposit    104641
Non Refund     14587
Refundable       162
Name: count, dtype: int64

In [23]:
# Calculate the cancellation rate for each deposit type
cancellation_by_deposit = (
    df.groupby('deposit_type')['is_canceled'].mean() * 100
)

cancellation_by_deposit.sort_values(ascending=False)

deposit_type
Non Refund    99.362446
No Deposit    28.377022
Refundable    22.222222
Name: is_canceled, dtype: float64

In [33]:
# Summarize total, canceled, and non-canceled bookings by deposit type
deposit_summary = df.groupby('deposit_type').agg(
    total_bookings=('is_canceled', 'count'),
    canceled_bookings=('is_canceled', 'sum')
)

# Calculate the number of non-canceled bookings
deposit_summary['non_canceled_bookings'] = (
    deposit_summary['total_bookings']
    - deposit_summary['canceled_bookings']
)

deposit_summary

,total_bookings,canceled_bookings,non_canceled_bookings
deposit_type,,,
No Deposit,104641,29694,74947
Non Refund,14587,14494,93
Refundable,162,36,126


### Cancellation Analysis by Deposit Type

Bookings with the Non Refund deposit type have the highest cancellation rate at approximately 99.36%, compared with 28.38% for No Deposit and 22.22% for Refundable bookings.

This is a strong association in the dataset, but it does not establish that the deposit type causes cancellations. The Refundable category contains only 162 bookings, so its cancellation rate should be interpreted cautiously.

In [25]:
# Count bookings for each hotel type
df['hotel'].value_counts()

hotel
City Hotel      79330
Resort Hotel    40060
Name: count, dtype: int64

In [26]:
# Calculate the cancellation rate for each hotel type
cancellation_by_hotel = (
    df.groupby('hotel')['is_canceled'].mean() * 100
)

cancellation_by_hotel.sort_values(ascending=False)

hotel
City Hotel      41.726963
Resort Hotel    27.763355
Name: is_canceled, dtype: float64

In [29]:
# Summarize bookings by hotel type
hotel_summary = df.groupby('hotel').agg(
    total_bookings=('is_canceled', 'count'),
    canceled_bookings=('is_canceled', 'sum')
)

# Calculate non-canceled bookings
hotel_summary['non_canceled_bookings'] = (
    hotel_summary['total_bookings']
    - hotel_summary['canceled_bookings']
)

hotel_summary

# Calculate the cancellation rate as a percentage
hotel_summary['cancellation_rate_%'] = (
    hotel_summary['canceled_bookings']
    / hotel_summary['total_bookings'] * 100
)

hotel_summary

,total_bookings,canceled_bookings,non_canceled_bookings,cancellation_rate_%
hotel,,,,
City Hotel,79330,33102,46228,41.726963
Resort Hotel,40060,11122,28938,27.763355


### Cancellation Analysis by Hotel Type

The City Hotel recorded 79,330 bookings, with 33,102 cancellations and a cancellation rate of 41.73%. The Resort Hotel recorded 40,060 bookings, with 11,122 cancellations and a cancellation rate of 27.76%.

The City Hotel's cancellation rate was approximately 13.96 percentage points higher than that of the Resort Hotel. This finding suggests that cancellation management may require greater attention at the City Hotel. However, further analysis is needed to investigate the factors associated with this difference.

In [34]:
# Count bookings for each customer type
df['customer_type'].value_counts()

customer_type
Transient          89613
Transient-Party    25124
Contract            4076
Group                577
Name: count, dtype: int64

In [35]:
# Calculate the cancellation rate for each customer type
cancellation_by_customer = (
    df.groupby('customer_type')['is_canceled'].mean() * 100
)

# Sort customer types by cancellation rate in descending order
cancellation_by_customer.sort_values(ascending=False)

customer_type
Transient          40.746320
Contract           30.961727
Transient-Party    25.429868
Group              10.225303
Name: is_canceled, dtype: float64

In [36]:
# Summarize bookings by customer type
customer_summary = df.groupby('customer_type').agg(
    total_bookings=('is_canceled', 'count'),
    canceled_bookings=('is_canceled', 'sum')
)

# Calculate non-canceled bookings
customer_summary['non_canceled_bookings'] = (
    customer_summary['total_bookings']
    - customer_summary['canceled_bookings']
)

# Calculate the cancellation rate as a percentage
customer_summary['cancellation_rate_%'] = (
    customer_summary['canceled_bookings']
    / customer_summary['total_bookings'] * 100
)

# Sort by total bookings in descending order
customer_summary = customer_summary.sort_values(
    'total_bookings', ascending=False
)

# Display the summary table
customer_summary

,total_bookings,canceled_bookings,non_canceled_bookings,cancellation_rate_%
customer_type,,,,
Transient,89613,36514,53099,40.746320
Transient-Party,25124,6389,18735,25.429868
Contract,4076,1262,2814,30.961727
Group,577,59,518,10.225303


### Cancellation Analysis by Customer Type

Transient bookings have the highest cancellation rate at 40.75% and the largest number of cancellations, with 36,514 canceled bookings. Contract bookings have a cancellation rate of 30.96%, while Transient-Party bookings have a rate of 25.43%. Group bookings show the lowest cancellation rate at 10.23%, based on 577 total bookings.

These findings suggest that cancellation patterns vary across customer types. However, further analysis is needed to understand the factors associated with these differences.

In [37]:
# Display descriptive statistics for lead time
df['lead_time'].describe()

count    119390.000000
mean        104.011416
std         106.863097
min           0.000000
25%          18.000000
50%          69.000000
75%         160.000000
max         737.000000
Name: lead_time, dtype: float64

In [39]:
# Group bookings by lead time
df['lead_time_group'] = pd.cut(
    df['lead_time'],
    bins=[-1, 7, 30, 90, float('inf')],
    labels=['0-7 days', '8-30 days', '31-90 days', 'Over 90 days']
)

# Count bookings in each lead time group
df['lead_time_group'].value_counts().sort_index()

lead_time_group
0-7 days        19746
8-30 days       18960
31-90 days      29553
Over 90 days    51131
Name: count, dtype: int64

In [40]:
# Analyze cancellations by lead time group
lead_time_summary = df.groupby('lead_time_group', observed=False).agg(
    total_bookings=('is_canceled', 'count'),
    canceled_bookings=('is_canceled', 'sum')
)

# Calculate non-canceled bookings
lead_time_summary['non_canceled_bookings'] = (
    lead_time_summary['total_bookings']
    - lead_time_summary['canceled_bookings']
)

# Calculate cancellation rate as a percentage
lead_time_summary['cancellation_rate_%'] = (
    lead_time_summary['canceled_bookings']
    / lead_time_summary['total_bookings'] * 100
)

# Display the summary table
lead_time_summary

,total_bookings,canceled_bookings,non_canceled_bookings,cancellation_rate_%
lead_time_group,,,,
0-7 days,19746,1902,17844,9.632331
8-30 days,18960,5283,13677,27.863924
31-90 days,29553,11141,18412,37.698372
Over 90 days,51131,25898,25233,50.650290


### Cancellation Analysis by Lead Time

The cancellation rate increases as the time between booking and arrival becomes longer. Bookings made 0–7 days before arrival have the lowest cancellation rate at 9.63%, while bookings made more than 90 days in advance have the highest rate at 50.65%.

Bookings made more than 90 days in advance also account for the largest number of cancellations, with 25,898 canceled bookings. These findings suggest that bookings made well in advance may require closer monitoring. However, this analysis identifies an association and does not establish that a longer lead time causes cancellations.

In [41]:
# Explore the average daily rate
df['adr'].describe()

count    119390.000000
mean        101.831122
std          50.535790
min          -6.380000
25%          69.290000
50%          94.575000
75%         126.000000
max        5400.000000
Name: adr, dtype: float64

In [42]:
# Explore the average daily rate
df['adr'].describe()

count    119390.000000
mean        101.831122
std          50.535790
min          -6.380000
25%          69.290000
50%          94.575000
75%         126.000000
max        5400.000000
Name: adr, dtype: float64

In [43]:
# Count negative ADR values
(df['adr'] < 0).sum()

1

In [44]:
# Display the highest ADR values
df['adr'].nlargest(10)

48515     5400.00
111403     510.00
15083      508.00
103912     451.50
13142      450.00
13391      437.00
39155      426.25
39568      402.00
39118      397.38
13323      392.00
Name: adr, dtype: float64

In [45]:
# Inspect the booking with the highest ADR
df.loc[df['adr'] == 5400]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,lead_time_group
48515,City Hotel,1,35,2016,March,13,25,0,1,2,...,12.0,NaN,0,Transient,5400.0,0,0,Canceled,2016-02-19,31-90 days


In [46]:
# Inspect the booking with the highest ADR
df.loc[df['adr'] == 5400, [
    'hotel',
    'is_canceled',
    'lead_time',
    'stays_in_weekend_nights',
    'stays_in_week_nights',
    'adults',
    'children',
    'babies',
    'adr',
    'reservation_status'
]]

,hotel,is_canceled,lead_time,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,adr,reservation_status
48515,City Hotel,1,35,0,1,2,0.0,0,5400.0,Canceled


In [47]:
# Count bookings with negative ADR
(df['adr'] < 0).sum()

1

In [48]:
# Inspect the booking with a negative ADR
df.loc[df['adr'] < 0, [
    'hotel',
    'is_canceled',
    'lead_time',
    'stays_in_weekend_nights',
    'stays_in_week_nights',
    'adults',
    'children',
    'babies',
    'adr',
    'reservation_status'
]]

,hotel,is_canceled,lead_time,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,adr,reservation_status
14969,Resort Hotel,0,195,4,6,2,0.0,0,-6.38,Check-Out


In [49]:
# Count bookings with ADR greater than 500
(df['adr'] > 500).sum()

3

In [50]:
# Inspect bookings with ADR greater than 500
df.loc[df['adr'] > 500, [
    'hotel',
    'is_canceled',
    'lead_time',
    'stays_in_weekend_nights',
    'stays_in_week_nights',
    'adults',
    'children',
    'babies',
    'adr',
    'reservation_status'
]]

,hotel,is_canceled,lead_time,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,adr,reservation_status
15083,Resort Hotel,0,1,0,1,2,0.0,0,508.0,Check-Out
48515,City Hotel,1,35,0,1,2,0.0,0,5400.0,Canceled
111403,City Hotel,0,0,0,1,1,0.0,0,510.0,Check-Out


In [51]:
# Count bookings with zero total nights
(
    df['stays_in_weekend_nights']
    + df['stays_in_week_nights']
    == 0
).sum()

715

In [52]:
# Analyze reservation status for bookings with zero nights
df.loc[
    (df['stays_in_weekend_nights'] + df['stays_in_week_nights']) == 0,
    'reservation_status'
].value_counts()

reservation_status
Check-Out    680
Canceled      22
No-Show       13
Name: count, dtype: int64

In [53]:
# Summarize ADR for bookings with zero nights
zero_night_bookings = df[
    (df['stays_in_weekend_nights'] + df['stays_in_week_nights']) == 0
]

zero_night_bookings['adr'].describe()

count    715.0
mean       0.0
std        0.0
min        0.0
25%        0.0
50%        0.0
75%        0.0
max        0.0
Name: adr, dtype: float64

In [54]:
# Calculate the interquartile range (IQR) for ADR
Q1 = df['adr'].quantile(0.25)
Q3 = df['adr'].quantile(0.75)
IQR = Q3 - Q1

# Calculate the lower and upper outlier boundaries
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Display the results
print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower bound:", lower_bound)
print("Upper bound:", upper_bound)

Q1: 69.29
Q3: 126.0
IQR: 56.709999999999994
Lower bound: -15.774999999999991
Upper bound: 211.065


In [55]:
# Count ADR outliers using the IQR boundaries
outliers = df[
    (df['adr'] < lower_bound) |
    (df['adr'] > upper_bound)
]

print("Number of ADR outliers:", len(outliers))

Number of ADR outliers: 3793


In [56]:
# Display the 10 highest ADR outliers
outliers.nlargest(10, 'adr')[[
    'hotel',
    'is_canceled',
    'stays_in_weekend_nights',
    'stays_in_week_nights',
    'adults',
    'children',
    'adr',
    'reservation_status'
]]

,hotel,is_canceled,stays_in_weekend_nights,stays_in_week_nights,adults,children,adr,reservation_status
48515,City Hotel,1,0,1,2,0.0,5400.00,Canceled
111403,City Hotel,0,0,1,1,0.0,510.00,Check-Out
15083,Resort Hotel,0,0,1,2,0.0,508.00,Check-Out
103912,City Hotel,0,1,1,2,2.0,451.50,Check-Out
13142,Resort Hotel,1,4,10,2,0.0,450.00,Canceled
13391,Resort Hotel,1,2,4,2,2.0,437.00,Canceled
39155,Resort Hotel,0,2,6,2,2.0,426.25,Check-Out
39568,Resort Hotel,0,2,3,3,1.0,402.00,Check-Out
39118,Resort Hotel,0,3,5,3,2.0,397.38,Check-Out
13323,Resort Hotel,1,2,8,2,1.0,392.00,Canceled


In [57]:
# Analyze ADR outliers by reservation status
outliers['reservation_status'].value_counts()

reservation_status
Check-Out    2324
Canceled     1446
No-Show        23
Name: count, dtype: int64

In [58]:
# Analyze ADR outliers by hotel
outliers['hotel'].value_counts()

hotel
Resort Hotel    2503
City Hotel      1290
Name: count, dtype: int64

In [60]:
# Calculate the percentage of ADR outliers by hotel
outlier_counts = outliers['hotel'].value_counts()
hotel_counts = df['hotel'].value_counts()

outlier_percentage = (outlier_counts / hotel_counts) * 100

outlier_percentage

hotel
City Hotel      1.626119
Resort Hotel    6.248128
Name: count, dtype: float64

In [61]:
# Compare ADR statistics by reservation status
df.groupby('reservation_status')['adr'].describe()

,count,mean,std,min,25%,50%,75%,max
reservation_status,,,,,,,,
Canceled,43017.0,105.205241,52.747471,0.00,72.9,96.30,127.83,5400.00
Check-Out,75166.0,99.987693,49.206263,-6.38,67.5,92.50,125.00,510.00
No-Show,1207.0,96.378459,45.030888,0.00,69.4,86.39,118.80,328.67


In [62]:
# Analyze ADR for completed stays
completed_stays = df[df['reservation_status'] == 'Check-Out']

completed_stays['adr'].describe()

count    75166.000000
mean        99.987693
std         49.206263
min         -6.380000
25%         67.500000
50%         92.500000
75%        125.000000
max        510.000000
Name: adr, dtype: float64

In [63]:
# Calculate the average ADR for completed stays
completed_adr_mean = completed_stays['adr'].mean()

# Exclude IQR outliers from completed stays for comparison
completed_adr_filtered = completed_stays[
    (completed_stays['adr'] >= lower_bound) &
    (completed_stays['adr'] <= upper_bound)
]

filtered_adr_mean = completed_adr_filtered['adr'].mean()

print("Average ADR before filtering:", completed_adr_mean)
print("Average ADR after excluding IQR outliers:", filtered_adr_mean)

Average ADR before filtering: 99.98769297288668
Average ADR after excluding IQR outliers: 95.28541308585707


In [64]:
# Compare ADR statistics by hotel type
df.groupby('hotel')['adr'].agg(
    ['count', 'mean', 'median', 'min', 'max']
)

,count,mean,median,min,max
hotel,,,,,
City Hotel,79330,105.304465,99.9,0.00,5400.0
Resort Hotel,40060,94.952930,75.0,-6.38,508.0


### ADR Analysis by Hotel Type

City Hotel has a higher average ADR of 105.30 compared with 94.95 for Resort Hotel. The median ADR is also higher for City Hotel (99.90) than for Resort Hotel (75.00), suggesting that the difference is not explained only by a few extreme values.

City Hotel has a maximum ADR of 5,400, which may substantially affect its average. Resort Hotel also contains a negative ADR value of -6.38, which requires further investigation.

These results show differences in recorded daily rates between the two hotel types. However, ADR alone does not measure total revenue or profitability.